odetoolbox/system_of_shapes.py

```diff
@@ -179,9 +188,21 @@ def generate_propagator_solver(self, output_timestep_symbol="__h"):
                     sym_str = "__P__{}__{}".format(str(self.x_[row]), str(self.x_[col]))
                     P_sym[row, col] = sympy.parsing.sympy_parser.parse_expr(sym_str, global_dict=Shape._sympy_globals)
                     P_expr[sym_str] = P[row, col]
-                    update_expr_terms.append(sym_str + " * " + str(self.x_[col]))
-            update_expr[str(self.x_[row])] = " + ".join(update_expr_terms) + " + " + str(self.c_[row])
-            update_expr[str(self.x_[row])] = sympy.simplify(sympy.parsing.sympy_parser.parse_expr(update_expr[str(self.x_[row])], global_dict=Shape._sympy_globals))
+                    if _is_zero(self.b_[row]):
+                        # homogeneous ODE
+                        update_expr_terms.append(sym_str + " * " + str(self.x_[col]))
+                    else:
+                        # inhomogeneous ODE
+                        # check that off-diagonal elements are zero in the system matrix to check that this is at most a first-order system
+                        for _col in range(P_sym.shape[1]):
+                            if _col != row:
+                                assert _is_zero(self.A_[row, _col]), "Higher-order inhomogeneous ODEs are not supported"
+
+                        particular_solution = -self.b_[row] / self.A_[row, row]
+                        update_expr_terms.append(sym_str + " * (" + str(self.x_[col]) + " - (" + str(particular_solution) + "))" + " + (" + str(particular_solution) + ")")
+                    assert _is_zero(self.c_[row]), "nonlinear part should be zero for propagators"
+            update_expr[str(self.x_[row])] = " + ".join(update_expr_terms)
+            update_expr[str(self.x_[row])] = sympy.parsing.sympy_parser.parse_expr(update_expr[str(self.x_[row])], global_dict=Shape._sympy_globals)
```

Comment:
Why is the `sympy.simplify` removed here?

Reply:
It's a summation of terms of the form c1*x + c2*x' + c3*x'', so there's probably nothing to be simplified. The inhomogeneous expressions might benefit from simplification, so I've added it back for them specifically.